.gitlab-ci.yml

```diff
@@ -81,69 +81,149 @@ build:nmodl:
     SPACK_PACKAGE_COMPILER: gcc
 
 # Build CoreNEURON
-build:coreneuron:mod2c:nvhpc:acc:
-  extends: [.build, .spack_nvhpc]
-  variables:
-    SPACK_PACKAGE: coreneuron
-    # See https://github.com/BlueBrain/CoreNeuron/issues/518 re: build_type
-    SPACK_PACKAGE_SPEC: +gpu+openmp+tests~legacy-unit build_type=RelWithDebInfo
-
-# Build CoreNEURON with Unified Memory on GPU
-build:coreneuron:mod2c:nvhpc:acc:unified:
-  extends: [.build, .spack_nvhpc]
-  variables:
-    SPACK_PACKAGE: coreneuron
-    # See https://github.com/BlueBrain/CoreNeuron/issues/518 re: build_type
-    SPACK_PACKAGE_SPEC: +gpu+unified+openmp+tests~legacy-unit build_type=RelWithDebInfo
-
-.build_coreneuron_nmodl:
+.build_coreneuron:
   extends: [.build]
   variables:
+    SPACK_PACKAGE: coreneuron
     # NEURON depends on py-mpi4py, most of whose dependencies are pulled in by
     # nmodl%gcc, with the exception of MPI, which is pulled in by
     # coreneuron%{nvhpc,intel}. hpe-mpi is an external package anyway, so
     # setting its compiler is just changing how it is labelled in the
     # dependency graph and not changing which installation is used, but this
     # means that in the NEURON step an existing py-mpi4py%gcc can be used.
     # Otherwise a new py-mpi4py with hpe-mpi%{nvhpc,intel} will be built.
+    # caliper: papi%nvhpc does not build; use the caliper from the deployment
     # TODO: fix this more robustly so we don't have to play so many games.
-    SPACK_PACKAGE_DEPENDENCIES: ^hpe-mpi%gcc
+    SPACK_PACKAGE_DEPENDENCIES: ^hpe-mpi%gcc ^caliper%gcc+cuda cuda_arch=70
+
+build:coreneuron:mod2c:nvhpc:acc:
+  extends: [.build_coreneuron, .spack_nvhpc]
+  variables:
+    SPACK_PACKAGE_SPEC: +caliper+gpu+openmp~shared+tests~legacy-unit build_type=RelWithDebInfo
+
+build:coreneuron:mod2c:nvhpc:acc:debug:
+  extends: [.build_coreneuron, .spack_nvhpc]
+  variables:
+    SPACK_PACKAGE_SPEC: +caliper+gpu+openmp~shared+tests~legacy-unit build_type=Debug
+
+# Shared + OpenACC + OpenMP host threading has problems
```

Comment:
See https://github.com/BlueBrain/CoreNeuron/issues/810. I think we also still have problems (not tested in the CI) with OpenMP offload and shared library execution.

Reply:
Ok.

Reply:
Given that our host threading is quite simple, I wonder if we might be better off using something simpler (`std::thread` or some thin layer on top) than OpenMP. We seem to run into a lot of problems mixing and matching OpenACC/OpenMP and OpenMP host/target...

Reply:
We can discuss/explore that but I think it would be also helpful to find out if the issue is programming models themselves or at the logical level in multi-threading+gpu interactions with neuron/coreneuron.